rfcs/0002-validation/text.md

```diff
@@ -0,0 +1,149 @@
+toolchains. We want to support as many of those toolchains as we can, so we're going to convert datasets into
+different formats. It seems to make sense to do this right after validation is complete since we're already parsing
+the original file anyway, and there's not really a good reason to delay converting.
+
+## Product Requirements
+
+The key requirement here is that users can prepare their own datasets without intervention from CZI. This goes along
+with all the other work to support self-submission: file upload, collection creation, and publication.
+
+## Detailed Design
+
+### Overview
+
+![Three Phases of Dataset Creation](imgs/three_phases.png)
+
+There are three main phases in the creation of a new dataset in the data portal:
+
+1. The submitter prepares a dataset file. This is done locally, outside the data portal, assisted by tooling we provide.
+   The outcome of this phase is a file in a format the data portal accepts that follows the schema.
+2. The submitter needs to then upload the dataset file to the portal. This involves things like creating a collection
+   and performing whatever consent/attestation is required. Then the submitter actually uploads the file, a process
+   which is described in detail in other RFCs.
+3. The uploaded file then needs to be validated and processed so it can be available via the portal. This involves four
+   main tasks:
+   - Validating that the file follows the schema.
+   - Updating the portal database with information about the uploaded file, things like tissue or ethnicity that come
+     from the schema.
+   - Converting the uploaded file to different formats and associating those with the dataset.
+   - Creating a cellxgene deployment for the dataset.
+
+This RFC is concerned with phase 3.
+
+### Relevant Components
+
+There are six main components involved in validation and processing:
+
+![High-level Validation Architecture](imgs/validation_architecture.png)
+
+1. Staging bucket. The S3 bucket where uploaded file are placed by the upload process.
+2. Artifact bucket. The S3 bucket where artifacts associated with datasets are stored and served.
+3. cellxgene deployment bucket. The S3 bucket configured to serve datasets via cellxgene.
+4. Portal database. The database where information about datasets is stored. The portal API queries this database when
+   responding to requests.
+5. Validation and processing image. The Docker image containing the dependencies and script to validate and process.
+6. Validation and processing Fargate task. The task definition for executing the image with AWS Fargate.
+
+#### Portal Database Schema
+
+The relevant part of the portal database is three tables: `dataset`, `artifacts`, and `deployment_directories`.
+
+![Database Schema](imgs/db_schema.png)
+
+The `dataset` table tracks the status of datasets and various metadata about them that is used by the portal frontend.
+There are three status fields:
+
+- upload\_status: uploading, uploaded, canceled
```

Comment:
Is there any reporting of  _progress_ in the [GET](https://github.com/chanzuckerberg/single-cell/blob/ff4dbd272b90ae0901b9b522c4c5c0f4252973ab/rfcs/RFC-Upload-From-Dropbox.md#get-submissionssubmission_idupload) response? (Step 1 of X completed or ...)

Reply:
@Bento007 I'm good with putting it all together in one field. I think that means "Failed" need to become "Upload Failed" so we can distinguish different types of failures. Then we'd have an Uploaded-->Validating-->Valid-->Creating Artifacts-->Artifacts Complete happy path right?

@brianraymor I think we include status in the GET, so the frontend will have something to display.

Reply:
I think that we need one of the RFCs (or the update to the high level arch doc) to be precise about the reporting states so it's easier for @tihuan and @hthomas-czi to close on the UX design.

Reply:
So the content of the status field can really be driven by whatever is best for UX. Looking at @hthomas-czi 's [quick design](https://www.figma.com/file/g0L0CfVzRc8r6HQa4NH1wJ/Collections?node-id=174%3A0), it looks like maybe we do want to split out the status a bit so it can be clear that a data is Valid _and_ had one error in conversion. What other information would you want to give to the user?

Reply:
Actually thinking a bit more, this is why I separated the three status fields. I was thinking the user and frontend are basically asking three questions:

1. Did the upload work?
2. Is my file valid?
3. Is it done processing and ready to go?

To answer (1) with separate fields, you can just do like `response["upload_status"] == "Complete"`. If it's one status field then the logic is more like `if response["status"] in ("Validating", "Valid", "Invalid", ...)` because you know those states are only reachable after upload is complete. That seems like a lot of backend logic to stick in the frontend.

In the alternative, if the question users are asking is more like "What's happening right now?", then the single status is better.

Reply:
0. How's the upload progressing? Should I go and have some coffee?
3. Why is processing hung? Or why have 2 out of 3 matrices appeared, but not the 3rd?